File: fcp/tensorflow/task_eligibility_info_ops.cc
/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <string>

#include "fcp/protos/federated_api.pb.h"
#include "tensorflow/core/framework/common_shape_fns.h"
#include "tensorflow/core/framework/dataset.h"
#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/op_kernel.h"
#include "tensorflow/core/framework/shape_inference.h"
#include "tensorflow/core/lib/core/errors.h"
#include "tensorflow/core/public/version.h"

namespace fcp {

using ::google::internal::federatedml::v2::TaskEligibilityInfo;
using ::google::internal::federatedml::v2::TaskWeight;

/**
 * CreateTaskEligibilityInfo op-kernel. Converts a set of input tensors into a
 * `TaskEligibilityInfo` proto serialized into a string tensor.
 *
 * This op is used to generate `TaskEligibilityInfo` protos from a model at
 * runtime, since TF Mobile does not support the standard TensorFlow ops for
 * encoding/decoding protos.
 */
class CreateTaskEligibilityInfoOp : public tensorflow::OpKernel {
 public:
  explicit CreateTaskEligibilityInfoOp(
      tensorflow::OpKernelConstruction* context)
      : OpKernel(context) {}

  void Compute(tensorflow::OpKernelContext* ctx) override {
    // Note: We use the tensorflow::data::ParseScalar/VectorArgument helpers
    // here, even though this op isn't strictly related to our tf.Dataset
    // integration. The helpers are public though, and we already use them in
    // our ExternalDataset implementation, so we might as well use them here
    // too.

    // Parse/validate the input arguments.
    tensorflow::int64 version;
    OP_REQUIRES_OK(
        ctx, tensorflow::data::ParseScalarArgument(ctx, "version", &version));
    std::vector<tensorflow::tstring> task_names;
    OP_REQUIRES_OK(ctx, tensorflow::data::ParseVectorArgument(ctx, "task_names",
                                                              &task_names));
    std::vector<float> task_weights;
    OP_REQUIRES_OK(ctx, tensorflow::data::ParseVectorArgument(
                            ctx, "task_weights", &task_weights));
    OP_REQUIRES(ctx, task_names.size() == task_weights.size(),
                tensorflow::errors::InvalidArgument(absl::StrCat(
                    "task_names length must match task_weights length: ",
                    task_names.size(), " vs. ", task_weights.size())));

    // Create the output proto, based on the inputs.
    TaskEligibilityInfo eligibility_info;
    eligibility_info.set_version(version);
    // Create a `TaskWeight` message for each pair of `task_names` and
    // `task_weights` elements.
    auto task_weight_it = task_weights.cbegin();
    for (const tensorflow::tstring& task_name : task_names) {
      float task_weight = *task_weight_it++;
      TaskWeight* task_weight_proto = eligibility_info.add_task_weights();
      task_weight_proto->set_task_name(std::string(task_name));
      task_weight_proto->set_weight(task_weight);
    }

    // Place the serialized output proto into the output tensor.
    tensorflow::Tensor* output_tensor;
    OP_REQUIRES_OK(ctx,
                   ctx->allocate_output("output", tensorflow::TensorShape({}),
                                        &output_tensor));
    output_tensor->scalar<tensorflow::tstring>()() =
        eligibility_info.SerializeAsString();
  }
};

REGISTER_OP("CreateTaskEligibilityInfo")
    .Input("version: int64")
    .Input("task_names: string")
    .Input("task_weights: float32")
    .Output("output: string")
    .SetShapeFn(tensorflow::shape_inference::ScalarShape);

REGISTER_KERNEL_BUILDER(
    Name("CreateTaskEligibilityInfo").Device(tensorflow::DEVICE_CPU),
    CreateTaskEligibilityInfoOp);

}  // namespace fcp

File: fcp/tensorflow/task_eligibility_info_ops.py
# Copyright 2021 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Ops for creating TaskEligibilityInfo results."""

import tensorflow as tf

# Ops implemented in C++
from fcp.tensorflow import gen_task_eligibility_info_ops

_task_eligibility_info_ops_so = tf.load_op_library(
    tf.compat.v1.resource_loader.get_path_to_datafile(
        "./_task_eligibility_info_ops.so"))


def create_task_eligibility_info(version, task_names, task_weights):
  """Outputs a serialized `TaskEligibilityInfo` proto based on the given inputs.

  This op is used to generate `TaskEligibilityInfo` protos from a model at
  runtime, since TF Mobile does not support the standard TensorFlow ops for
  encoding/decoding protos.

  See the `TaskEligibilityInfo` and `TaskWeight` proto message documentation for
  more information.

  Args:
    version: an int64 value to place in the `TaskEligibilityInfo.version` field.
    task_names: a rank-1 string tensor containing the task names to assign
      weights to. Each entry in this tensor will be combined with the
      corresponding entry into the `task_weights` tensor at the same index, to
      form a `TaskWeight` message.
    task_weights: a rank-1 float tensor containing the task weight for each task
      (see `task_names`). Note: this tensor must have the same number of
        elements as `task_names`.

  Returns:
    a string tensor containing the serialized proto.
  """
  # Convert the inputs to tensors, as a convenience to callers. This ensures
  # that they can easily pass regular Python or numpy types in addition to
  # actual tensors.
  version = tf.convert_to_tensor(version, dtype=tf.int64, name="version")
  task_names = tf.convert_to_tensor(
      task_names, dtype=tf.string, name="task_names")
  task_weights = tf.convert_to_tensor(
      task_weights, dtype=tf.float32, name="task_weights")
  return gen_task_eligibility_info_ops.create_task_eligibility_info(
      version=version, task_names=task_names, task_weights=task_weights)
